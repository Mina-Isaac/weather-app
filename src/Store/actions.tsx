import { createAction } from "typesafe-actions";
import { actionTypes, weatherSegment } from "../constatnts";


const getData = createAction(actionTypes.FETCH_DATA);

const setData = createAction(actionTypes.FETCH_DATA_SUCCESS,
  resolve => (data: Array<weatherSegment>) =>
    resolve(data)
);

const dataError = createAction(actionTypes.FETCH_DATA_ERROR,
  resolve => (error: Error) =>
    resolve(error)
);

const setTempScale = createAction(actionTypes.SET_TEMP_SCALE,
  resolve => (scale :string) =>
  resolve(scale)
);

export {
  getData,
  setData,
  dataError,
  setTempScale
};
