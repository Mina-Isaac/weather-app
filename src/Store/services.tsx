import { API_URL, APPID, city, segmentCount } from "../constatnts";


export class appServices {
  static fetchDataFromAPI() {
    const url = `${API_URL + city}&APPID=${APPID}&cnt=${segmentCount}`;
    return fetch(url).then((res) => res.json().then((data) => data));
  }
}
