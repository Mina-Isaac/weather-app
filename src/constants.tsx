export const API_URL = 'http://api.openweathermap.org/data/2.5/forecast?q=';
export const APPID = '75f972b80e26f14fe6c920aa6a85ad57'
export const city = 'Munich,de'
export const segmentCount= '40'
export const ICON_URL = 'http://openweathermap.org/img/wn/'


export const ActionTypes = {
  SET_TEMP_SCALE: 'SET_TEMP_SCALE',
  FETCH_DATA: 'FETCH_DATA',
  FETCH_DATA_SUCCESS: 'FETCH_DATA_SUCCESS',
  FETCH_DATA_ERROR: 'FETCH_DATA_ERROR',
};

export enum Scales {
  Fahrenheit = 'F',
  Celsius = 'C'

}

type main = {
  temp: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  sea_level: number;
  grnd_level: number;
  humidity: number;
  temp_kf: number;
}

type weather = {
  id: number;
  main: string;
  description: string;
  icon: string;
}

export type weatherSegment = {
  dt: number;
  main: main;
  weather: weather[];
  dt_txt: string
}
