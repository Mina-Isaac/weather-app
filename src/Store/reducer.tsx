import { ActionType, getType } from "typesafe-actions";
import * as actions from "./actions";
import { weatherSegment, Scales } from "../constants";

export type appState = {
  groupedSegments: weatherSegment[][];
  fetching: boolean;
  tempScale: string;
  selectedDay: number | undefined;
};

export type Action = ActionType<typeof actions>;

export const initialState: appState = {
  groupedSegments: [],
  fetching: true,
  tempScale: Scales.Fahrenheit,
  selectedDay: undefined
};

const reducer = (state: appState = initialState, action: any): appState => {
  switch (action.type) {
    case getType(actions.setData):
      return {
        fetching: false,
        groupedSegments: action.payload,
        tempScale: state.tempScale,
        selectedDay: state.selectedDay
      };

    case getType(actions.dataError):
      return {
        fetching: false,
        groupedSegments: [],
        tempScale: state.tempScale,
        selectedDay: state.selectedDay
      };

    case getType(actions.setTempScale):
      return { ...state, tempScale: action.payload };

    case getType(actions.setSelectedDay):
      return { ...state, selectedDay: action.payload };
    default:
      return state;
  }
};

export default reducer;
