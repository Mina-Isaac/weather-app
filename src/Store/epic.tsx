import { of, from } from "rxjs";
import { Epic } from "redux-observable";
import * as actions from "./actions";
import {
  catchError,
  map,
  filter,
  switchMap,
  withLatestFrom
} from "rxjs/operators";
import { ActionType, isActionOf } from "typesafe-actions";
import { appState } from "./reducer";


type Action = ActionType<typeof actions>;
const propertyFilter = ['dt', 'main', 'weather', 'dt_txt']

export const fetchWeatherSegmentsFlow: Epic<Action, Action, appState, any> = (
  action$,
  state$,
  { service }
) =>
  action$.pipe(
    filter(isActionOf(actions.getData)),
    withLatestFrom(state$),
    switchMap(arr =>
      from(service.fetchDataFromAPI()).pipe(
        map((data: any) => {
          const list = data.list;
          const filtered = list.map((item: any) => {
            const holder: any = {}
            for (let prop in item) {
              if (propertyFilter.includes(prop)) holder[prop] = item[prop]
            }
            return holder
          })
          return filtered
        }),
        map(actions.setData),
        catchError(error => of(actions.dataError(error)))
      )
    )
  );
