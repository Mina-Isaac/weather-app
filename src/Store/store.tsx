import {
  applyMiddleware,
  compose,
  createStore,
  Middleware,
} from "redux";
import { createEpicMiddleware } from "redux-observable";
import { fetchWeatherSegmentsFlow } from "./epic";
import { ActionType } from "typesafe-actions";
import { appServices } from "./services";
import * as actions from './actions'
import reducer, { appState } from './reducer'

type Action = ActionType<typeof actions>;


const epicMiddleware = createEpicMiddleware<Action, Action, appState>({
  dependencies: { service: appServices }
});
const middleware: [Middleware] = [epicMiddleware];

const configStore = () => {
  const createStoreWithMiddleware = compose(applyMiddleware(...middleware))(
    createStore
  );
  const store = createStoreWithMiddleware(reducer);
  epicMiddleware.run(fetchWeatherSegmentsFlow);

  return {
    store
  };
};
const { store } = configStore();
export { store };
