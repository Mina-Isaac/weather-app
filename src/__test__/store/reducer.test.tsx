import reducer, { appState } from '../../Store/reducer';
import * as actions from "../../Store/actions";
import { weatherSegment, Scales } from '../../constants';

export const mockWeatherSegment: weatherSegment = {
  dt: 0,
  main: {
    temp: 0,
    temp_min: 0,
    temp_max: 0,
    pressure: 0,
    sea_level: 0,
    grnd_level: 0,
    humidity: 0,
    temp_kf: 0,
  },
  weather: [
    {
      id: 0,
      main: "",
      description: "",
      icon: "10d"
    }
  ],
  dt_txt: ''
}
const getInitialState = (initial?: Partial<appState>) =>
reducer(initial as appState, {} as any);

describe('Reducer data flow', () => {
  describe('initial state', () => {
    it('should match a snapshot', () => {
      const initialState = getInitialState();
      expect(initialState).toMatchSnapshot();
    });
  });

  it('should put retrieved data in the store and set fetching flag to false with action setData', () => {
    const initialState = getInitialState();
    initialState.fetching = true;
    expect(initialState.segments).toHaveLength(0);
    const state = reducer(initialState, actions.setData([mockWeatherSegment]));
    expect(state.segments).toHaveLength(1);
    expect(state.fetching).toBeFalsy();
  });
  it('should set segment property to be [] and set fetching flag to false with action dataError', () => {
    const initialState = getInitialState();
    initialState.fetching = true;
    initialState.segments = [mockWeatherSegment];
    const state = reducer(initialState, actions.dataError((new Error('some error occurred'))));
    expect(state.segments).toHaveLength(0);
    expect(state.fetching).toBeFalsy();
  });
  it('should have Fahrenheit as default scale and set temperature scale with action setTempScale', () => {
    const initialState = getInitialState();
    expect(initialState.tempScale).toBe(Scales.Fahrenheit);
    const state = reducer(initialState, actions.setTempScale(Scales.Celsius));
    expect(state.tempScale).toBe(Scales.Celsius);
  });
  it('should have undefined as the default selectedDay and set it with action setSelectedDay', () => {
    const initialState = getInitialState();
    expect(initialState.selectedDay).toBe(undefined);
    const selectedDay=4;
    const state = reducer(initialState, actions.setSelectedDay(4));
    expect(state.selectedDay).toBe(selectedDay);
  });
});
