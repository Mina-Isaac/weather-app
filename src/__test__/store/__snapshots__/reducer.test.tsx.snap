// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer data flow initial state should match a snapshot 1`] = `
Object {
  "fetching": true,
  "segments": Array [],
  "selectedDay": undefined,
  "tempScale": "F",
}
`;
